### ISSUE TITLE
Unable to download full data set

### ISSUE THREAD

**Issue by IanMiddlebrook** [BiologicalRecordsCentre/UKBMS-online#364]
Hi @DavidRoy @Gary-van-Breda 

I've been trying all morning to download the full 2024 data from the Annual Summary page - filters set to 'Full data Set (All Users)' and 'Combine data for all sites together'.

I've had no problem with most downloads, but each time I try to retrieve the Occurrences download, I am greeted with this message after about 3 minutes:

**Service Unavailable**
The server is temporarily unable to service your request due to maintenance downtime or capacity problems. Please try again later.

Strangely, I was able to download the Section Level table first time, and this is usually a slightly bigger file than the Occurrences.

But the Occurrences table is critical for building a database to help validate the 2024 data.

Many thanks, Ian

**Comment by IanMiddlebrook** [comment]
Hi @DavidRoy 
Could I bump this one please?
Inability to download the full 'Occurrences' could impact the data validation timeline.

Many thanks, Ian

**Comment by DavidRoy** [comment]
@Gary-van-Breda @andrewvanbreda can you pick this up please

**Comment by andrewvanbreda** [comment]
@DavidRoy 
I will take a look

**Comment by andrewvanbreda** [comment]
Hi @DavidRoy @IanMiddlebrook ,

The trouble appears to be that the report is getting slower further into the year as more records come in, hence we are having trouble in November. Eventually the report is becoming so slow the website is timing out.
What the report is doing is quite ambitious which is making it slow, and actually I think it is difficult to make it faster.
The report is a couple of years old, and hasn't been changed since it was created, so I think it is the amount of data that is breaking it (it actually works for some other years).

I can see a couple of optimisations I could do to the report, but I don't think these will be enough to make it fast enough.
I would be interested to know if any of the columns being returned are not necessary?

The current column selection is this:

Samples ID
Occurrence ID
Site Type
Site Code
Date
Section No.
Species
Preferred
Abundance Count

What I have done is this, run the report for a year with not many records, so that the report actually works. 
I have then been able to collect the SQL from the logs and I can then run the SQL directly against the database and change the year to 2024. In this way I can get the data you need before any changes are made to improve the report.

Ian: I will mail you the data you wanted. The data seems to be 40+ megabytes, and unfortunately I don't have a way to compress it on the system that is running the database so I will have to split the data up.
You should be able to combine it into 1 file at your end.

Please let me know if any of those columns are not needed, and I will see if I can use that information to speed up the report.

Thanks

Andy

**Comment by IanMiddlebrook** [comment]
Thanks @andrewvanbreda 
This table works in conjunction with the Samples download, which already includes the 'Site Type' and 'Site Code' columns.
So those values will already be associated with each Sample ID, and could be removed from this download.

**Comment by andrewvanbreda** [comment]
Hi @IanMiddlebrook,

I have sent you six files via 6 separate emails, I hope they arrive OK.
I forgot to mention that I included the header row in all 6 files in case you don't want to combine them, if you do want to combine them then that row will needing removing from all but the first file.

Thanks for the info about the report, I will see if I can speed it up.

**Comment by IanMiddlebrook** [comment]
Thanks  for the files @andrewvanbreda 

I can make some progress with those now.
Unfortunately I do usually need to take the download several times during the validation process, as more data are added and corrections are made. So hopefully a sustainable solution can be found.
Thanks, Ian

**Comment by andrewvanbreda** [comment]
Hi @IanMiddlebrook 
OK thanks for letting me know about the urgency regarding this.

**Comment by DavidRoy** [comment]
@andrewvanbreda @johnvanbreda @JimBacon is there an option to run a scheduled task to create the download files. So it runs in the background and can be given a longer time-out

**Comment by Gary-van-Breda** [comment]
Possible tweaks to the report:

- Only add in the joins for “samples parent” and “locations l” as part of the param "user_id" (Ian has said we can ditch the l.code, we can replace all the parent dates and survey_id with the equivalents on the cache_occurrences_functional (o) table. The location type term is actually part of the file name – on top of the fact that Ian has said we can ditch that too – so we can also ditch the join to “cache_termlists_terms ctt”)
- Ditch the “samples child” join, and replace the child.location_id with the o.location_id (on the cache_occurrences_functional table)

This means together that for a full data dump on the year (for all users) we can reduce the report from a 7 table join to 3 tables…

**Comment by Gary-van-Breda** [comment]
Update - I didn't consider the location filtering properly:

- Ditch the l.code column
- Ditch the location_type_id param, as sufficiently covered by the survey_id
- Replace the l.code sort by with parent.location_id
- Ditch the "locations l" Join: will need to ditch the locattrs and loc2attrs params
- Ditch the "samples child" join (child.location_id replaced with o.location_id)
- Ditch the site_type column, and the "cache_termlists_terms ctt" Join

This reduces it to a 4 table join.
It may be worth testing if the date and survey filters (under the parameters) would provide better performance if done against the cache_occurrences_functional fields rather than the parent sample.

**Comment by Gary-van-Breda** [comment]
@IanMiddlebrook : is it OK to remove the site code "sort by"?

**Comment by IanMiddlebrook** [comment]
@Gary-van-Breda  sure, doesn't really matter what order the data are in.

**Comment by andrewvanbreda** [comment]
@Gary-van-Breda 
OK thanks for the thoughts, some of those match up with what I was thinking, some new ideas.
I have have a test on Warehouse1 pgadmin, then commit to report.

@DavidRoy 
I guess the SQL could be run as a scheduled task, but I am not sure about what we would do with any output. Any data generated would need to be accessible to the right people, if it saved to a file on the Warehouse I am not sure how we would give appropriate access. I will try some of Gary's suggestions, he has a mixture there of things I was thinking of doing, but some new ideas too so a scheduled task might not be needed anyway.

**Comment by johnvanbreda** [comment]
@DavidRoy we could add new code to the warehouse which does something like the following:
1. Add a new parameter to the reporting system - asyncReporting=true UKBMS can be configured to add to the parameters for this report.
2. When a report request is detected that has the asyncReporting parameter set, then the reporting API creates a work queue task of a new type "task_run_async_report". It saves the supplied parameters and the user's email address as parameters for the work queue task entry.
3. When the work queue picks up this task later, it runs the report using the supplied parameters, compresses the result, places it in the same location as Elasticsearch report downloads get saved, then sends a link to the email address.

There may still be problems if the PostgreSQL database is configured to limit queries so they time out at 5 minutes, which it used to do (though I'm not sure if that still applies since the cloud server move).

There's a day or so's work here to get this working (assuming the timeouts aren't a problem).

It would be good to follow through on Andy & Gary's efforts to optimise the report first anyway.

**Comment by andrewvanbreda** [comment]
I have created a new test version of the report that is only available on my UKBMS pantheon multidev at the moment, it is a lot faster (maybe 30 seconds to a minute to run) and does download.
There is just one issue at the minute, that actually it is returning a few thousand fewer records than I had  expected, and there is no obvious reason for this looking at the SQL differences (unless location_id in cache_occurrences_functional isn't behaving in the way we assumed).
Unfortunately my jumpbox to Warehouse1 is broken today, so once IT Support have fixed that I will able to debug.
But I am now hopeful we will be able to get the speed we need without resorting to complex solutions.

@johnvanbreda 
Note, I thought an open link such as an email link wouldn't be suitable in this case as this particular report is designed to include confidential records.

**Comment by IanMiddlebrook** [comment]
@andrewvanbreda 
Could the difference in the number of records be due to missing out 'confidential' sites?
These sites are currently not presenting in the Summary Table drop-down list, although the data were certainly in the Samples download - see issue #365

**Comment by andrewvanbreda** [comment]
@IanMiddlebrook 
That is good point, and is seems more likely than anything else I can think of to be honest. I will have a check of the file output and see if it is those that are missing.

**Comment by andrewvanbreda** [comment]
Hi @IanMiddlebrook ,

Ian, you were right that is was the confidential records that were missing.

I have adjusted my test version of the report so that it now returns those, and include as many of the suggested optimisations as possible (minus the ones that broke the confidential records returning).
It now seems to download the file ok.

The new report is available here if you want to try it
https://avb-iform-brc-ukbms-d10.pantheonsite.io/

This site is a clone of live UKBMS with a few differences, but you will find the download in the same place as before.

Before we get to the stage of updating the live version of the report, I want to clear up some oddities that I noticed.

1. The new report is returning more records than the one I sent you a few days ago Ian. I noticed some records which are appearing and aren't new in past few days.
I cannot be sure this is a comprehensive list without my database access (which isn't sorted out yet), but I believe the following occurrences are in a situation where the occurrence is attached to a sub-sample which is in the "UKBMS Transects" survey, but the parent sample is in "iRecord general data".
This resulted in them not appearing in the previous report but appearing in the new version.
Does anyone (@Gary-van-Breda @DavidRoy ) have any explanation for this? In particular, is it ok for these to appear in the report. Personally I think it is probably better for a report to show odd data rather than leave it invisible.

These are the occurrence IDs

38704490
38704488
38704487
38704482
38704481
38704479
38704478
38704473
38704467
38704466
38704465
38704463
38704412
38704393
38704392

2. The following are also only appearing in the new report but not the old one, but actually I have no idea why. I don't think I can tell until I can check the database cache tables, but be aware of their existence. 

36025592
36025591
36025590
36025589
36025588
36025586

4. I have also attached a list of occurrences that appeared in the old report but not in the new. I checked a few of them manually and I think these are probably all deleted records.
It may simply be because they have been deleted since I sent you that data Ian, but I can't tell until I get my db access back.
They are attached in a file for reference anyway.

Anyway in summary.
I would be interested in anyone's thoughts on point 1.
I would probably like to get an answer to 2 and 3 before putting the new report live but I probably need to wait for my database access.

[Likely deleted occs.csv](https://github.com/user-attachments/files/17778179/Likely.deleted.occs.csv)

**Comment by IanMiddlebrook** [comment]
Thanks @andrewvanbreda 
Point 1 - the sample_id related to these occurrences does not appear in the samples download. In which case they are of no value, as we do not know the site location or conditions under which the data were collected.
Point 2 - strange one - these occurrences seem to be related to a site location which has been deleted (possibly a duplicate site?). I'm not sure how the records persisted, but I've checked the recorder's account, and can see which location they were intended for, so have amended the entry.
Point 3 - I can confirm these occurrences relate to entries that have recently been deleted.

Best wishes, Ian

**Comment by andrewvanbreda** [comment]
Hi @IanMiddlebrook 

1. I can see the report can be changed to exclude these without affecting performance.
I will change the report so the samples/sub-samples/occurrences must all be from the same survey.
2. I can confirm samples that have location deleted will persist. UKBMS doesn't do anything special regarding this, and standard behaviour is the sample won't be deleted. I wasn't present for any discussions regarding this originally, but I would guess this is at least partially because we can't assume any samples/occurrences with a removed location are invalid for a specific project (e.g. someone might remove the location and then add the location as a free text name after)
3. Great

I will test the report change from point 1 and let you know when the new version of the report is on the live site.

**Comment by andrewvanbreda** [comment]
Hi @IanMiddlebrook,

This issue is all fixed and live now.
Close issue if happy, or comment further.

Thanks

Andy
